Greeting service repository and tests

BrainDias opened the pull request.

BrainDias committed:
Little changes

BrainDias committed:
Changed Id in User entity to jakarta package

BrainDias committed:
Added lombok to build.gradle

BrainDias committed:
1. Added annotationProcessor dependency to build.gradle
2. Injected mocked repo into GreetingServiceImpl

BrainDias committed:
Verified using of UserRepository method

ahlinist commented on build.gradle:
```diff
@@ -20,7 +20,11 @@ ext {
 	implementation 'org.springframework.boot:spring-boot-starter-web'
 	implementation 'org.springframework.cloud:spring-cloud-starter-openfeign'
-	testImplementation 'org.springframework.boot:spring-boot-starter-test'
+    implementation 'jakarta.persistence:jakarta.persistence-api:3.1.0'
```
indentation

ahlinist commented on build.gradle:
```diff
@@ -20,7 +20,11 @@ ext {
+	implementation 'org.springframework.boot:spring-boot-starter-data-jpa:3.0.4'
+	implementation 'org.projectlombok:lombok:1.18.26'
+	annotationProcessor 'org.projectlombok:lombok:1.18.26'
+    testImplementation 'org.springframework.boot:spring-boot-starter-test'
```
indentation

ahlinist commented on src/main/java/com/example/demo/Application.java:
```diff
@@ -6,7 +6,6 @@
 @SpringBootApplication
 public class Application {
     public static void main(String[] args) {
-        //ApplicationContext context = new AnnotationConfigApplicationContext("com.example.demo.services");
```
why aren't these changes made in a separate PR? This one is about greeting repository.

ahlinist commented on src/main/java/com/example/demo/Application.java:
```diff
@@ -6,7 +6,6 @@
-        //ApplicationContext context = new AnnotationConfigApplicationContext("com.example.demo.services");
         SpringApplication.run(Application.class, args);
     }
 }
```
blank line is better to be added in the end

ahlinist commented on src/test/java/com/example/demo/controller/GreetingControllerTests.java:
```diff
@@ -2,14 +2,16 @@
 import static org.mockito.Mockito.when;
 
+@ExtendWith(MockitoExtension.class)
 class GreetingControllerTests {
```
what has controller to do with a repository? These changes are supposed to be added in a separate PR.

ahlinist commented on src/test/java/com/example/demo/controller/GreetingControllerTests.java:
```diff
@@ -18,7 +20,7 @@ class GreetingControllerTests {
-        when(greetingService.printHello(name)).thenReturn("Hello Mary!");
-        assertEquals(greetingController.hello(name), "Hello Mary!");
+        when(greetingService.printHello(name)).thenReturn("Hello Mrs. Mary!");
+        assertEquals(greetingController.hello(name), "Hello Mrs. Mary!");
```
where are mock verifications?

BrainDias committed:
Merge branch 'main' into GreetingService_repository_and_tests

# Conflicts:
#	build.gradle
#	src/test/java/com/example/demo/controllers/GreetingControllerTests.java

BrainDias committed:
New branch for SecurityConfig

ahlinist commented on src/main/java/com/example/demo/repository/UserRepository.java:
```diff
@@ -0,0 +1,9 @@
+
+public interface UserRepository extends CrudRepository<User,Long> {
+
+    User findUserByName(String name);
```
Consider using optional here

ahlinist commented on src/main/java/com/example/demo/services/GreetingServiceImpl.java:
```diff
@@ -1,11 +1,16 @@
+
     public String printHello(String name) {
-        return "Hello " + name + "!";
+        return "Hello " + userRepository.findUserByName(name).getTitle() + name + "!";
```
what happens if no user is found?

BrainDias committed:
Created Object entity model

BrainDias committed:
Created user persistence and handling basics

BrainDias committed:
Updated object entity model with ProductOrder and Auction

BrainDias committed:
Changed Product and Weapon to abstract classes and made them @Embeddable,deleted redundant Accessory class

BrainDias committed:
Renamed repository package to repositories

BrainDias committed:
Changed type of Product @ElementCollection

BrainDias committed:
Added API for products requests

BrainDias committed:
Added Auction functionality and mailing for closed auctions and submitted orders

BrainDias committed:
Covered main services with unit tests

ahlinist commented on src/main/java/com/example/demo/Application.java:
```diff
@@ -2,6 +2,9 @@
 
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.scheduling.annotation.Scheduled;
```
unused imports?

ahlinist commented on src/main/java/com/example/demo/controllers/AuctionController.java:
```diff
@@ -0,0 +1,28 @@
+
+@RestController
+@RequiredArgsConstructor
+public class AuctionController {
```
You should add @RequestMapping("/auction") here

ahlinist commented on src/main/java/com/example/demo/controllers/AuctionController.java:
```diff
@@ -0,0 +1,28 @@
+public class AuctionController {
+    private final AuctionService auctionService;
+
+    @PostMapping("/auction")
```
remove annotation argument from here

ahlinist commented on src/main/java/com/example/demo/controllers/AuctionController.java:
```diff
@@ -0,0 +1,28 @@
+        auctionService.createAuction(owner,dto,startPrice,title);
+    }
+
+    @PutMapping("/bid/{auction_id}")
```
The URL is supposed to be /auction/<id>/bid

So the annotation should look like: @PutMapping("{id}/bid")

ahlinist commented on src/main/java/com/example/demo/controllers/AuctionController.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    @PutMapping("/bid/{auction_id}")
+    public HttpStatusCode placeBidToAuction(@AuthenticationPrincipal User pretender,@RequestParam Float price,@PathVariable Long auction_id){
```
No request (query) params should be in PUT.

Poor naming for auction_id: first of all this violates any Java code conventions, should be camelCase. Second: why not calling it "id"?

Could auth be moved away from controller? Any filters or something implicit?

ahlinist commented on src/main/java/com/example/demo/controllers/AuctionController.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    @PutMapping("/bid/{auction_id}")
+    public HttpStatusCode placeBidToAuction(@AuthenticationPrincipal User pretender,@RequestParam Float price,@PathVariable Long auction_id){
```
Please don't return status code as a return type/value.

Status codes are controller by respective spring annotations. In your case this should be 201 or 204. By default spring returns 200, so you should set status explicitly.

BrainDias replied:
Used service method returns 404, 412 or 202 depending on inside logic. So there is no default/always-returning status. What exactly is the problem?

ahlinist commented on src/main/java/com/example/demo/controllers/AuctionController.java:
```diff
@@ -0,0 +1,28 @@
+
+    @PutMapping("/bid/{auction_id}")
+    public HttpStatusCode placeBidToAuction(@AuthenticationPrincipal User pretender,@RequestParam Float price,@PathVariable Long auction_id){
+        return auctionService.placeBidToAuction(pretender,price,auction_id);
```
No principals should go to business layer.

ahlinist commented on src/main/java/com/example/demo/controllers/AuctionController.java:
```diff
@@ -0,0 +1,28 @@
+    private final AuctionService auctionService;
+
+    @PostMapping("/auction")
+    public void createAuction(@AuthenticationPrincipal User owner,
```
this will return status 200 on success, but in your case the return status might be 201 or 204

ahlinist replied:
no request params should be passed for POST, auth is better to be handled elsewhere

ahlinist commented on src/main/java/com/example/demo/controllers/AuctionController.java:
```diff
@@ -0,0 +1,28 @@
+                                        @RequestBody AuctionDTO dto,
+                                        @RequestParam Float startPrice,
+                                        @RequestParam String title){
+        auctionService.createAuction(owner,dto,startPrice,title);
```
no principals should go to business layer

ahlinist commented on src/main/java/com/example/demo/controllers/OrderController.java:
```diff
@@ -0,0 +1,72 @@
+import org.springframework.http.HttpStatusCode;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.core.annotation.AuthenticationPrincipal;
+import org.springframework.web.bind.annotation.*;
```
please avoid wildcard imports, import classes explicitly

ahlinist commented on src/main/java/com/example/demo/controllers/AuctionController.java:
```diff
@@ -0,0 +1,28 @@
+                                        @RequestBody AuctionDTO dto,
+                                        @RequestParam Float startPrice,
+                                        @RequestParam String title){
+        auctionService.createAuction(owner,dto,startPrice,title);
```
no dto's should go into business layer, they need to be mapped to models or entities in the controller

BrainDias replied:
What in case if mapping of DTO depends on the state of the entity stored in DB? (e.g. type)

ahlinist commented on src/main/java/com/example/demo/entities/Auction.java:
```diff
@@ -0,0 +1,50 @@
+package com.example.demo.entities;
+
+import com.example.demo.products.Product;
+import jakarta.persistence.*;
```
wildcard import

ahlinist commented on src/main/java/com/example/demo/entities/Auction.java:
```diff
@@ -0,0 +1,50 @@
+
+    @CreatedDate
+    @Temporal(TemporalType.TIMESTAMP)
+    private Date createdAt;
```
use Instant instead

ahlinist commented on src/main/java/com/example/demo/entities/Auction.java:
```diff
@@ -0,0 +1,50 @@
+
+    @LastModifiedDate
+    @Temporal(TemporalType.TIMESTAMP)
+    private Date updatedAt;
```
and here

ahlinist commented on src/main/java/com/example/demo/entities/Auction.java:
```diff
@@ -0,0 +1,50 @@
+    private Date updatedAt;
+
+    @Temporal(TemporalType.TIMESTAMP)
+    private Date closingAt;
```
also here

ahlinist commented on src/main/java/com/example/demo/entities/ImageEntity.java:
```diff
@@ -0,0 +1,15 @@
+package com.example.demo.entities;
+
+import jakarta.persistence.*;
```
wildcard import

BrainDias replied:
IJ IDEA automatically imports wildcard when using @Lob. Seems to be it's bug I haven't noticed

ahlinist commented on src/main/java/com/example/demo/entities/ImageEntity.java:
```diff
@@ -0,0 +1,15 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Lob
+    private byte[] imageData;
```
well, in general lobs/blobs are not quite recommended in RDBMS

BrainDias replied:
What to use instead?

ahlinist commented on src/main/java/com/example/demo/entities/ProductOrder.java:
```diff
@@ -0,0 +1,43 @@
+    private List<Product> products;
+
+    Float price;
+    boolean delivered;
```
use wrappers in entities: Boolean

ahlinist commented on src/main/java/com/example/demo/entities/ProductOrder.java:
```diff
@@ -0,0 +1,43 @@
+
+    @CreatedDate
+    @Temporal(TemporalType.TIMESTAMP)
+    private Date createdAt;
```
instant

ahlinist commented on src/main/java/com/example/demo/entities/ProductOrder.java:
```diff
@@ -0,0 +1,43 @@
+
+    @LastModifiedDate
+    @Temporal(TemporalType.TIMESTAMP)
+    private Date updatedAt;
```
instant

ahlinist commented on src/main/java/com/example/demo/entities/User.java:
```diff
@@ -0,0 +1,80 @@
+
+//    @Email
+    private String email;
+    private boolean active;
```
no primitives in entities

ahlinist commented on src/main/java/com/example/demo/entities/User.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Override
+    public boolean isAccountNonExpired() {
+        return true;
```
hardcoded trues?

BrainDias replied:
Yes, while we have no logic for account expiration. But UserDetails interface requires this method.

ahlinist commented on src/main/java/com/example/demo/products/AccessoryType.java:
```diff
@@ -0,0 +1,5 @@
+package com.example.demo.products;
+
+public enum AccessoryType {
+    Light, Muzzle_Device, Grip
```
enum values are supposed to be underscored uppercased: MUZZLE_DEVICE - like that

ahlinist commented on src/main/java/com/example/demo/products/AutoFire.java:
```diff
@@ -0,0 +1,5 @@
+package com.example.demo.products;
+
+public interface AutoFire {
+    Integer rateOfFire = null;
```
it is already null, without explicitly stating that

why would you use fields in interfaces? a lil cringy, isn't it?

ahlinist commented on src/main/java/com/example/demo/products/Caliber.java:
```diff
@@ -0,0 +1,31 @@
+
+public enum Caliber {
+    AK_47("7.62x39"),
+    RifleNATO("7.62x51"),
```
poor case, AK_47 above is much nicer

ahlinist commented on src/main/java/com/example/demo/products/FeedingType.java:
```diff
@@ -0,0 +1,5 @@
+package com.example.demo.products;
+
+public enum FeedingType {
+    Belt,Drum,Boxed
```
case

ahlinist commented on src/main/java/com/example/demo/products/Product.java:
```diff
@@ -0,0 +1,15 @@
+public abstract class Product {
+    public Float price;
+
+    public boolean forSale;
```
primitives

ahlinist commented on src/main/java/com/example/demo/repositories/AuctionRepository.java:
```diff
@@ -0,0 +1,10 @@
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface AuctionRepository extends PagingAndSortingRepository<Auction,Long>, CrudRepository<Auction,Long> {
```
why extending CrudRepo? doesn't PagingAndSorting have everything you need?

BrainDias replied:
PagingAndSorting includes only two findAll methods for Page's and Sort's. I will remove PagingAndSorting in repositories which don't use it in future

ahlinist commented on src/main/java/com/example/demo/services/AuctionService.java:
```diff
@@ -0,0 +1,79 @@
+    private final AuctionRepository auctionRepository;
+    private final UserRepository userRepository;
+
+    @Transactional
```
why @Transactional?

BrainDias replied:
Two WRITE queries into DB and they must be executed transactionally. In case it is not transactional there might be inconsistency if saving new auction will rollback while products selected for new auction are already deleted from this user.

ahlinist commented on src/main/java/com/example/demo/services/AuctionService.java:
```diff
@@ -0,0 +1,79 @@
+        userRepository.save(owner);
+        Auction auction = new Auction();
+        auction.setClosed(false);
+        auction.setOwner(owner);
```
should owner and auction be saved in cascade?

BrainDias replied:
Are there any possible alternatives where it is guaranteed that we won't lose owner products?

ahlinist commented on src/main/java/com/example/demo/services/AuctionService.java:
```diff
@@ -0,0 +1,79 @@
+        List<Product> products = dto.getProducts();
+        owner.getProducts().removeAll(products);
+        userRepository.save(owner);
+        Auction auction = new Auction();
```
wouldn't it be nicer to pass those params into constructor?

BrainDias replied:
now it works with Auction type parameter passed to the method (mapped from DTO in controller method) and we are only changing 2 fields in it

ahlinist commented on src/main/java/com/example/demo/services/AuctionService.java:
```diff
@@ -0,0 +1,79 @@
+    }
+    @Scheduled(fixedDelay = 1,timeUnit = TimeUnit.MINUTES)
+    public void closeAuctions(){
+        Streamable<Auction> auctions = (Streamable<Auction>) auctionRepository.findAll();
```
No, just update the auctions in the DB, don't load them all into memory.

you might want to use JPQL.

ahlinist commented on src/main/java/com/example/demo/services/AuctionService.java:
```diff
@@ -0,0 +1,79 @@
+    @Scheduled(fixedDelay = 1,timeUnit = TimeUnit.MINUTES)
+    public void closeAuctions(){
+        Streamable<Auction> auctions = (Streamable<Auction>) auctionRepository.findAll();
+        Streamable<Auction> auctionsToClose = auctions.filter(auction -> auction.getClosingAt().before(new Date()));
```
you've created new Date() n times

ahlinist commented on src/main/java/com/example/demo/services/AuctionService.java:
```diff
@@ -0,0 +1,79 @@
+
+@Service
+@RequiredArgsConstructor
+public class AuctionService {
```
why doesn't it implement an interface?

BrainDias replied:
What interface it should implement?

ahlinist commented on src/main/java/com/example/demo/services/AuctionService.java:
```diff
@@ -0,0 +1,79 @@
+            auction.setClosed(true);
+            userRepository.save(pretender);
+            try {
+                notifyAuctionClosed(auction.getOwner(),pretender);
```
OMG, super conunter-intuitive and implicit stuff. Just inject an instance of notification service.

ahlinist commented on src/main/java/com/example/demo/services/AuctionService.java:
```diff
@@ -0,0 +1,79 @@
+                throw new RuntimeException(e);
+            }
+        });
+        auctionRepository.saveAll(auctionsToClose);
```
any chance to cascade-save here?

BrainDias replied:
What do you mean by that?

ahlinist commented on src/main/java/com/example/demo/services/AuctionService.java:
```diff
@@ -0,0 +1,79 @@
+        auctionRepository.saveAll(auctionsToClose);
+    }
+
+    public HttpStatusCode placeBidToAuction(User pretender, Float price, Long auction_id) {
```
NOOOO! no http-stuff in services! Use them in controllers only!

BrainDias replied:
What in case if status to be returned is resolved inside internal service logic?

ahlinist commented on src/main/java/com/example/demo/services/AuctionService.java:
```diff
@@ -0,0 +1,79 @@
+
+    public HttpStatusCode placeBidToAuction(User pretender, Float price, Long auction_id) {
+        Optional<Auction> optionalAuction = auctionRepository.findById(auction_id);
+        if(optionalAuction.isEmpty()) return HttpStatus.BAD_REQUEST;
```
first of all: don't use ifs without code blocks ({})

second: use orElse

ahlinist commented on src/main/java/com/example/demo/services/AuctionService.java:
```diff
@@ -0,0 +1,79 @@
+        Optional<Auction> optionalAuction = auctionRepository.findById(auction_id);
+        if(optionalAuction.isEmpty()) return HttpStatus.BAD_REQUEST;
+        Auction auction = optionalAuction.get();
+        if(auction.getLastPrice()>=price||auction.isClosed()) return HttpStatus.PRECONDITION_FAILED;
```
code blocks and whitespaces between operators

ahlinist commented on src/main/java/com/example/demo/services/AuctionService.java:
```diff
@@ -0,0 +1,79 @@
+    public HttpStatusCode placeBidToAuction(User pretender, Float price, Long auction_id) {
+        Optional<Auction> optionalAuction = auctionRepository.findById(auction_id);
+        if(optionalAuction.isEmpty()) return HttpStatus.BAD_REQUEST;
+        Auction auction = optionalAuction.get();
```
should not be optional#get and then a bunch of setters, should be update inside repository

ahlinist commented on src/main/java/com/example/demo/services/EmailService.java:
```diff
@@ -0,0 +1,27 @@
+
+@Service
+public class EmailService {
+    @Autowired
```
use constructor injection instead

ahlinist commented on src/main/java/com/example/demo/services/EmailService.java:
```diff
@@ -0,0 +1,27 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class EmailService {
```
interface is missing

BrainDias replied:
Which one?

ahlinist commented on src/main/java/com/example/demo/services/NotificationService.java:
```diff
@@ -0,0 +1,35 @@
+    static String sellerText;
+
+    public static void notifyAuctionClosed(User owner, User pretender) throws MessagingException {
+        sendEmail(owner.getEmail(), ownerSubject,ownerText);
```
NOOOOOO! WHYYYYY!?!??!

ahlinist commented on src/main/java/com/example/demo/services/NotificationService.java:
```diff
@@ -0,0 +1,35 @@
+    @Value("${orer.seller.subject}")
+    static String sellerSubject;
+    @Value("${order.seller.text}")
+    static String sellerText;
```
does @Value work with static?

Why not using regular non-static fields and constructor injection? Or constants?

BrainDias committed:
Fixed issues with:
1. Fetching and filtering instead of SQL expressions
2. Using primitives in entities
3. case of enums and variables
4. mapping URLs in controllers
5. Request params in non-idempotent requests
6. Wildcard imports
7. Transferring DTOs to business later (partially)
8. Interfaces including fields only
9. Static fields and methods

BrainDias committed:
1. Added more functionality to auctions and orders
2. Changed product types structure and processing

BrainDias committed:
1. Added messages
2. Rewrote tests
3. Separated UserService and ProductService
4. Added filtering for products
5. Added ratings for users
6. Added phone numbers

BrainDias committed:
Merge remote-tracking branch 'refs/remotes/origin/Security_configs' into GreetingService_repository_and_tests

BrainDias committed:
1.Renamed project to Weaponplace
2.Removed productToDto from mapper
3. Deleted GreetingService.java
4. Optimized imports

BrainDias committed:
1.Renamed project to Weaponplace
2.Removed productToDto from mapper
3. Deleted GreetingService.java
4. Optimized imports
5. Added basic SecurityConfig with
BCryptPasswordEncoder, inMemoryUserDetailsService and SecurityFilterChain

BrainDias committed:
1. Added posgresql driver dependency to build.gradle
2. Configured datasource in application.yml

BrainDias committed:
1. Fixed bugs in migration file
2. New version of spring boot, flyway
3. Made Spring initialize MapStruct implementations as beans

BrainDias committed:
1. Configured to launch on port 8080
2. Changed name of AuctionController common path to "/auctions" to be more common with other controllers
3. Updated SecurityConfig to match Spring Security 6.1 standarts.
4. Enabled method security and specified it on /users/admin endpoint

BrainDias commented on src/main/java/org/weaponplace/services/implementations/UserServiceImpl.java:
```diff
@@ -0,0 +1,60 @@
+
+
+    public Page<User> pageUsers(Pageable pageRequest){
+        return userRepository.findAll(pageRequest);
```
fadfaddddddddddddddddddddddddddddddddddd

BrainDias committed:
Major changes lol :D
